File: infra/cdk.out/asset.0e1da2e40b6682900f6fd0cccea0f513a84372274770c0059783120dd21b8cf0/sources/football.py
"""
SignedData CDS — Football Ingestor
Source: api-football.com (RapidAPI free plan: 100 req/day)
Docs:   https://www.api-football.com/documentation-v3

Usage:
    ingestor = FootballIngestor(
        signer=signer,
        api_key="YOUR_RAPIDAPI_KEY",
        league_ids=[71],   # 71 = Brasileirão Série A
        season=2026,
    )
    events = await ingestor.ingest()
"""
from __future__ import annotations

import hashlib
from datetime import datetime, timezone
from typing import Any

import httpx

from cds.ingestor import BaseIngestor
from cds.schema import CDSEvent, ContextMeta, SourceMeta
from cds.sources.football_models import (
    FootballContentTypes,
    FootballMatchPayload,
    FootballTeam,
    FootballVenue,
)

API_BASE = "https://v3.football.api-sports.io"

LEAGUE_IDS: dict[str, int] = {
    "brasileirao_a":    71,
    "brasileirao_b":    72,
    "copa_brasil":      73,
    "libertadores":     13,
    "sul_americana":    11,
    "premier_league":   39,
    "champions_league": 2,
}

STATUS_MAP: dict[str, str] = {
    "NS": "scheduled",
    "1H": "live", "HT": "live", "2H": "live",
    "ET": "live", "BT": "live", "P": "live",
    "FT": "finished", "AET": "finished", "PEN": "finished",
    "PST": "postponed", "CANC": "cancelled", "ABD": "cancelled",
    "AWD": "finished", "WO": "finished",
}


def _summary(
    home: str, away: str,
    hs: int | None, as_: int | None,
    status: str, competition: str,
    minute: int | None,
) -> str:
    if status == "finished":
        return f"{competition}: {home} {hs} x {as_} {away} (finished)"
    if status == "live":
        m = f" {minute}'" if minute else ""
        return f"{competition}: {home} {hs} x {as_} {away} — LIVE{m}"
    if status == "scheduled":
        return f"{competition}: {home} vs {away} — scheduled"
    return f"{competition}: {home} vs {away} ({status})"


def _team(t: dict[str, Any]) -> FootballTeam:
    return FootballTeam(
        id=t.get("id"),
        name=t.get("name", ""),
        short_name=(t.get("name") or "")[:3].upper(),
        logo_url=t.get("logo", ""),
    )


class FootballIngestor(BaseIngestor):
    content_type = FootballContentTypes.MATCH_RESULT

    def __init__(
        self,
        signer: Any,
        api_key: str,
        league_ids: list[int] | None = None,
        season: int | None = None,
    ) -> None:
        super().__init__(signer)
        self.api_key    = api_key
        self.league_ids = league_ids or [LEAGUE_IDS["brasileirao_a"]]
        self.season     = season or datetime.now(timezone.utc).year

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        async with httpx.AsyncClient(timeout=15) as client:
            for league_id in self.league_ids:
                resp = await client.get(
                    f"{API_BASE}/fixtures",
                    headers={"x-apisports-key": self.api_key},
                    params={"league": league_id, "season": self.season, "last": 10},
                )
                resp.raise_for_status()
                fp = "sha256:" + hashlib.sha256(resp.content).hexdigest()
                for fixture in resp.json().get("response", []):
                    events.append(self._build_event(fixture, fp))
        return events

    def _build_event(self, f: dict[str, Any], fingerprint: str) -> CDSEvent:
        fx     = f["fixture"]
        teams  = f["teams"]
        goals  = f["goals"]
        league = f["league"]
        venue  = fx.get("venue", {})

        hs     = goals.get("home")
        as_    = goals.get("away")
        status = STATUS_MAP.get(fx.get("status", {}).get("short", "NS"), "scheduled")
        minute = fx.get("status", {}).get("elapsed")

        home_team       = _team(teams["home"]); home_team.score = hs
        away_team       = _team(teams["away"]); away_team.score = as_
        competition     = league.get("name", "")

        ct = (
            FootballContentTypes.MATCH_LIVE
            if status == "live"
            else FootballContentTypes.MATCH_RESULT
        )

        payload = FootballMatchPayload(
            match_id=fx.get("id"),
            home=home_team, away=away_team,
            status=status,  # type: ignore[arg-type]
            competition=competition,
            competition_id=league.get("id"),
            season=str(league.get("season", self.season)),
            matchday=(league.get("round") or "").replace("Regular Season - ", "") or None,
            venue=FootballVenue(name=venue.get("name", ""), city=venue.get("city", "")),
            minute=minute,
            referee=fx.get("referee", ""),
            match_date=(fx.get("date") or "")[:10],
        )

        return CDSEvent(
            content_type=ct,
            source=SourceMeta(id="api-football.com.v3", fingerprint=fingerprint),
            occurred_at=datetime.fromisoformat(
                fx["date"].replace("Z", "+00:00")
            ) if fx.get("date") else datetime.now(timezone.utc),
            lang="en",
            payload=payload.model_dump(mode="json"),
            context=ContextMeta(
                summary=_summary(home_team.name, away_team.name, hs, as_, status, competition, minute),
                model="rule-based-v1",
            ),
        )

File: sdk/typescript/src/sources/football.ts
/**
 * SignedData CDS — Football Source
 * Source: api-football.com (RapidAPI)
 */

import { createHash } from "node:crypto";
import { CDSEvent, ContextMeta, SourceMeta } from "../schema.js";
import type { CDSSigner } from "../signer.js";
import { BaseIngestor } from "../ingestor.js";
import { CDSVocab, CDSSources } from "../vocab.js";

export const FootballContentTypes = {
  MATCH_RESULT: CDSVocab.FOOTBALL_MATCH_RESULT,
  MATCH_LIVE:   CDSVocab.FOOTBALL_MATCH_LIVE,
  STANDINGS:    CDSVocab.FOOTBALL_STANDINGS,
} as const;

export const LEAGUE_IDS = {
  brasileirao_a: 71, brasileirao_b: 72, copa_brasil: 73,
  libertadores: 13, sul_americana: 11,
  premier_league: 39, champions_league: 2,
} as const;

export type MatchStatus = "scheduled" | "live" | "finished" | "postponed" | "cancelled";

export interface FootballTeam {
  id?: number; name: string; short_name: string;
  score: number | null; logo_url: string;
}
export interface FootballVenue { name: string; city: string; country: string; }
export interface FootballMatchPayload {
  match_id?: number; home: FootballTeam; away: FootballTeam;
  status: MatchStatus; competition: string; competition_id?: number;
  season: string; matchday?: string; venue: FootballVenue;
  minute?: number; referee: string; match_date: string;
}
export interface StandingsEntry {
  position: number; team: FootballTeam;
  played: number; won: number; drawn: number; lost: number;
  goals_for: number; goals_against: number; goal_diff: number;
  points: number; form: string;
}
export interface FootballStandingsPayload {
  competition: string; competition_id?: number; season: string;
  table: StandingsEntry[];
}

const STATUS_MAP: Record<string, MatchStatus> = {
  NS: "scheduled",
  "1H": "live", HT: "live", "2H": "live", ET: "live", BT: "live", P: "live",
  FT: "finished", AET: "finished", PEN: "finished",
  PST: "postponed", CANC: "cancelled", ABD: "cancelled",
  AWD: "finished", WO: "finished",
};

const API_BASE = "https://v3.football.api-sports.io";
type R = Record<string, unknown>;

function buildTeam(t: R, score: number | null): FootballTeam {
  return {
    id: t["id"] as number | undefined,
    name: (t["name"] as string) ?? "",
    short_name: ((t["name"] as string) ?? "").slice(0, 3).toUpperCase(),
    score, logo_url: (t["logo"] as string) ?? "",
  };
}

function summary(
  home: string, away: string, hs: number | null, as_: number | null,
  status: MatchStatus, comp: string, minute?: number
): string {
  if (status === "finished") return `${comp}: ${home} ${hs} x ${as_} ${away} (finished)`;
  if (status === "live")     return `${comp}: ${home} ${hs} x ${as_} ${away} — LIVE${minute ? ` ${minute}'` : ""}`;
  if (status === "scheduled") return `${comp}: ${home} vs ${away} — scheduled`;
  return `${comp}: ${home} vs ${away} (${status})`;
}

export interface FootballIngestorOptions {
  apiKey: string;
  leagueIds?: number[];
  season?: number;
}

export class FootballIngestor extends BaseIngestor {
  readonly contentType = FootballContentTypes.MATCH_RESULT;
  private readonly apiKey: string;
  private readonly leagueIds: number[];
  private readonly season: number;

  constructor(signer: CDSSigner, opts: FootballIngestorOptions) {
    super(signer);
    this.apiKey    = opts.apiKey;
    this.leagueIds = opts.leagueIds ?? [LEAGUE_IDS.brasileirao_a];
    this.season    = opts.season ?? new Date().getFullYear();
  }

  async fetch(): Promise<CDSEvent[]> {
    const events: CDSEvent[] = [];
    for (const leagueId of this.leagueIds) {
      const url  = `${API_BASE}/fixtures?league=${leagueId}&season=${this.season}&last=10`;
      const resp = await globalThis.fetch(url, { headers: { "x-apisports-key": this.apiKey } });
      if (!resp.ok) throw new Error(`api-football HTTP ${resp.status}`);
      const buf  = Buffer.from(await resp.arrayBuffer());
      const fp   = "sha256:" + createHash("sha256").update(buf).digest("hex");
      const body = JSON.parse(buf.toString("utf-8")) as { response: R[] };
      for (const f of body.response ?? []) events.push(this.buildEvent(f, fp));
    }
    return events;
  }

  private buildEvent(f: R, fingerprint: string): CDSEvent {
    const fx    = (f["fixture"] as R) ?? {};
    const teams = (f["teams"]   as Record<string, R>) ?? {};
    const goals = (f["goals"]   as Record<string, number | null>) ?? {};
    const lg    = (f["league"]  as R) ?? {};
    const venue = ((fx["venue"] as R) ?? {}) as Record<string, string>;
    const fxSt  = (fx["status"] as Record<string, unknown>) ?? {};

    const hs     = goals["home"] ?? null;
    const as_    = goals["away"] ?? null;
    const status: MatchStatus = STATUS_MAP[(fxSt["short"] as string) ?? "NS"] ?? "scheduled";
    const minute = fxSt["elapsed"] as number | undefined;

    const homeTeam = buildTeam(teams["home"] ?? {}, hs);
    const awayTeam = buildTeam(teams["away"] ?? {}, as_);
    const comp     = (lg["name"] as string) ?? "";
    const ct       = status === "live" ? FootballContentTypes.MATCH_LIVE : FootballContentTypes.MATCH_RESULT;

    const payload: FootballMatchPayload = {
      match_id: fx["id"] as number | undefined,
      home: homeTeam, away: awayTeam, status, competition: comp,
      competition_id: lg["id"] as number | undefined,
      season: String(lg["season"] ?? this.season),
      matchday: (lg["round"] as string)?.replace("Regular Season - ", "") || undefined,
      venue: { name: venue["name"] ?? "", city: venue["city"] ?? "", country: "" },
      minute, referee: (fx["referee"] as string) ?? "",
      match_date: ((fx["date"] as string) ?? "").slice(0, 10),
    };

    return new CDSEvent({
      content_type: ct,
      source: { "@id": CDSSources.API_FOOTBALL, fingerprint },
      occurred_at: (fx["date"] as string) ?? new Date().toISOString(),
      lang: "en",
      payload: payload as unknown as R,
      context: {
        summary: summary(homeTeam.name, awayTeam.name, hs, as_, status, comp, minute),
        model: "rule-based-v1",
        generated_at: new Date().toISOString(),
      },
    });
  }
}
